benchmarks/benches/complicated_maud.rs

```diff
@@ -0,0 +1,102 @@
+#![feature(plugin, test)]
+#![plugin(maud_macros)]
+
+extern crate maud;
+extern crate test;
+
+use maud::{Markup, PreEscaped};
+
+#[derive(Debug)]
+struct Entry {
+    name: &'static str,
+    score: u16,
+}
+
+mod btn {
+    use std::fmt::{self, Display, Formatter};
+
+    #[derive(Copy, Clone)]
+    pub enum RequestMethod {
+        Get, Post
+    }
+
+    #[derive(Copy, Clone)]
+    pub struct Button<'a> {
+        label: &'a str,
+        path: &'a str,
+        req_meth: RequestMethod,
+    }
+
+    impl<'a> Button<'a> {
+        pub fn new(label: &'a str, path: &'a str) -> Button<'a> {
+            Button {
+                label: label,
+                path: path,
+                req_meth: RequestMethod::Get,
+            }
+        }
+
+        pub fn with_method(mut self, meth: RequestMethod) -> Button<'a> {
+            self.req_meth = meth;
+            self
+        }
+    }
+
+    impl<'a> Display for Button<'a> {
+        fn fmt(&self, mut f: &mut Formatter) -> Result<(), fmt::Error> {
+            match self.req_meth {
+                RequestMethod::Get => {
+                    f.write_str(&html!(
+                        a.btn href=(self.path) (self.label)
+                    ).into_string())
+                }
+                RequestMethod::Post => {
+                    f.write_str(&html!(
+                        form method="POST" action=(self.path) {
+                            input.btn type="submit" value=(self.label) /
+                        }
+                    ).into_string())
+                }
+            }
+        }
+    }
+}
+
+fn layout<S: AsRef<str>>(title: S, inner: Markup) -> Markup {
```

Comment:
The use of `AsRef<str>` here is pretty clever! I wonder if it's a good idea to apply in general.